docs/src/userguide/faq.rst

```diff
@@ -0,0 +1,1018 @@
+
+.. code:: python
+
+    cdef char* s = "an unusual \0 containing C byte string"
+    a_python_byte_string = s[:21]    #  take the first 21 bytes of the string, including the \0 byte
+
+Note that this will not handle the case that the specified slice length is longer than the actual C string. This code will crash if the allocated memory area of the ``char*`` is shorter.
+
+There is also support for decoding a C string slice efficiently into a Python unicode string. Just do this:
+
+.. code:: python
+
+    # -*- coding: ISO8859-15
+    cdef char* s = "a UTF-8 encoded C string with fünny chäräctörs"
+    cdef Py_ssize_t byte_length = 46
+
+    a_python_unicode_string = s[:byte_length].decode('ISO8859-15')
+
+----------
+
+How do I pass a Python string parameter on to a C library?
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+See the `string tutorial<string_tutorial>`_.
+
+**Answer**: It depends on the semantics of the string. Imagine you have this C function:
+
+::
+
+    cdef extern from "something.h":
+        cdef int c_handle_data(char* data, int length)
+
+For __binary data__, you can simply require byte strings at the API level, so that this will work:
+
+::
+
+    def work_with_binary_data(bytes binary_data):
+        c_handle_data(binary_data, len(binary_data))
+
+It will raise an error (with a message that may or may not be appropriate for your use case) if users pass other things than a byte string.
+
+For __textual data__, however, you must handle Unicode data input. What you do with it depends on what your C function accepts. For example, if it requires UTF-8 encoded byte sequences, this might work:
+
+::
+
+    def work_with_text_data(text):
+        if not isinstance(text, unicode):
+            raise ValueError("requires text input, got %s" % type(text))
+        utf8_data = text.encode('UTF-8')
+        c_handle_data( utf8_data, len(utf8_data) )
+
+Note that this also accepts subtypes of the Python unicode type. Typing the "text" parameter as "unicode" will not cover this case.
+
+The above is the right thing to do in Py3. However, some (not all, just some) module APIs may become more user friendly in Python 2.x if you additionally allow well defined byte strings. For example, it may make sense to allow plain ASCII strings in some cases, as they are often used for textual data in Python 2.x programs. This could be done as follows:
+
+::
+
+    from python_version cimport PY_MAJOR_VERSION
+
+    def work_with_text_data(text):
+        if isinstance(text, unicode): # most common case first
+            utf8_data = text.encode('UTF-8')
+        elif (PY_MAJOR_VERSION < 3) and isinstance(text, str):
+            text.decode('ASCII') # trial decoding, or however you want to check for plain ASCII data
+            utf8_data = text
+        else:
+            raise ValueError("requires text input, got %s" % type(text))
+        c_handle_data(utf8_data, len(utf8_data))
```

Comment:
Py2 only, remove?

Reply:
Yeah, I'd remove this. It's for people who write new code for new APIs, and new code is rather unlikely to be written with Py2 in mind.